code/software/gpio-interrupt/gpio-interrupt.c

```diff
@@ -0,0 +1,135 @@
+/*
+ * vim: set et ts=2 sw=2
+ *------------------------------------------------------------
+ *                                  ___ ___ _   
+ *  ___ ___ ___ ___ ___       _____|  _| . | |_ 
+ * |  _| . |_ -|  _| . |     |     | . | . | '_|
+ * |_| |___|___|___|___|_____|_|_|_|___|___|_,_| 
+ *                     |_____|                       
+ * ------------------------------------------------------------
+ * Copyright (c) 2020 Xark
+ * MIT License
+ * ------------------------------------------------------------
+ */
+
+#include <basicio.h>
+#include <machine.h>
+#include <gpio.h>
+
+// macro for easier MFP register access (read or write)
+#define MFP(reg) (*(volatile uint8_t *)MFP_ ## reg)
+
+// rosco_m68k 100Hz loop_counter.  The MFP "timer C" interrupt handler
+// in rosco's firmware increments this every 100th of a second.
+volatile uint32_t *tick_100Hz = (uint32_t *)0x40C;
+
+// Bit 6 in IREB, IPRB and ISRB corresponds to rosco GPIO3
+// See page 3-4 in 1984 MC68901 MFP pdf
+#define MFP_GPIP4_BIT 0x40
+
+volatile int gpio3_interrupts;    // counter for number of GPIO3 interrupts
+
+// GPIO3 interrupt handler.  The "interrupt" attribute "tells the compiler to save
+// all the registers in this function and use the "rte" return from exception
+// instruction at the end instead of the normal "rts" return from subroutine instruction.
+// This allows the interrupt handler to be written in C.
+void gpio3_handler(void) __attribute__ ((interrupt));
+void gpio3_handler(void)
+{
+  MFP(GPDR) ^= 2;               // toggle red LED
+  gpio3_interrupts += 1;        // increment interrupt counter
+  print("<Interrupted!>");      // "interrupt" main printing
+  MFP(ISRB) = ~MFP_GPIP4_BIT;   // clear interrupt in-service bit
+}
+
+noreturn void kmain()
+{
+  // clear BSS segment as C standard requires (so bss variables start zero'd)
```

Comment:
Hmm, the `start_serial` lib should do this for you. If it isn't, it's a bug in the lib.

Reply:
I just removed that (again) and confirmed my "int gpio3_interrupts" (last bss in my test) starts at (e.g.) 0x813573e (IIRC :sweat_smile: ).  I will leave this out and assume you will apply "start_serial" OB1 fix (which I suspect, because it is very much like my first try at clearing which "didn't work" either).

Reply:
Somewhat interestingly, this didn't turn out to be off-by-one - it was a linkage issue. When I first looked I missed that we weren't talking about non-static globals - these go into `COMMON`, not `.BSS` - it's up to the link script to place `COMMON` correctly in the bss section, and my link scripts weren't doing that. So, `COMMON` was defaulting to being placed at the end of the `BSS` section, _after_ the `.bss_end` marker. Doing the less-than-or-equal to was making it clear to `bss_end`, and then carrying on into the COMMON.

TL;DR - I think I've fixed this. Please give it a go :)